pkg/volume/util/util_test.go

```diff
@@ -1460,6 +1460,612 @@ func TestSelectZoneForVolume(t *testing.T) {
+			ReplicaCount: 2,
+			AllowedTopologies: []v1.TopologySelectorTerm{
+				{
+					MatchLabelExpressions: []v1.TopologySelectorLabelRequirement{
+						{
+							Key:    kubeletapis.LabelZoneFailureDomain,
+							Values: []string{"zoneX"},
+						},
+					},
+				},
+			},
+			Reject: true,
+		},
+
+		// POSITIVE TESTS
+
+		// Select zones from zones parameter [Pass]
+		// [1] nil Node (Node irrelevant)
+		// [2] Zones parameter specified
+		// [3] no AllowedTopologies
+		// [4] ReplicaCount specified
+		{
+			Name:          "Zones_parameter_Dual_replica_Superset",
+			ZonesPresent:  true,
+			Zones:         "zoneW,zoneX,zoneY,zoneZ",
+			ReplicaCount:  2,
+			Reject:        false,
+			ExpectedZones: "zoneW,zoneX,zoneY,zoneZ",
+		},
+
+		// Select zones from zones parameter [Pass]
+		// [1] nil Node (Node irrelevant)
+		// [2] Zones parameter specified
+		// [3] no AllowedTopologies
+		// [4] ReplicaCount specified
+		{
+			Name:          "Zones_parameter_Dual_replica_Match",
+			ZonesPresent:  true,
+			Zones:         "zoneX,zoneY",
+			ReplicaCount:  2,
+			Reject:        false,
+			ExpectSpecificZones: true,
+			ExpectedZones: "zoneX,zoneY",
+		},
+
+		// Select zones from active zones [Pass]
+		// [1] nil Node (Node irrelevant)
+		// [2] no Zone parameter
+		// [3] no AllowedTopologies
+		// [4] ReplicaCount specified
+		{
+			Name:           "Active_zones_Nil_node_Dual_replica_Superset",
+			ZonesWithNodes: "zoneW,zoneX,zoneY,zoneZ",
+			ReplicaCount:   2,
+			Reject:         false,
+			ExpectedZones:  "zoneW,zoneX,zoneY,zoneZ",
+		},
+
+		// Select zones from active zones [Pass]
+		// [1] nil Node (Node irrelevant)
+		// [2] no Zone[s] parameter
+		// [3] no AllowedTopologies
+		// [4] ReplicaCount specified
+		{
+			Name:                "Active_zones_Nil_node_Dual_replica_Match",
+			ZonesWithNodes:      "zoneW,zoneX",
+			ReplicaCount:        2,
+			Reject:              false,
+			ExpectSpecificZones: true,
+			ExpectedZones:       "zoneW,zoneX",
+		},
+
+		// Select zones from node label and active zones [Pass]
+		// [1] Node with zone labels
+		// [2] no Zone parameter
+		// [3] no AllowedTopologies
+		// [4] ReplicaCount specified
+		{
+			Name:           "Active_zones_Node_with_Zone_labels_Dual_replica_Superset",
+			Node: nodeWithZoneLabels,
+			ZonesWithNodes: "zoneW,zoneX,zoneY,zoneZ",
+			ReplicaCount:   2,
+			Reject:         false,
+			ExpectSpecificZone: true,
+			ExpectedZone: "zoneX",
+			ExpectedZones:  "zoneW,zoneX,zoneY,zoneZ",
+		},
+
+		// Select zones from node label and active zones [Pass]
+		// [1] Node with zone labels
+		// [2] no Zone[s] parameter
+		// [3] no AllowedTopologies
+		// [4] ReplicaCount specified
+		{
+			Name:                "Active_zones_Node_with_Zone_labels_Dual_replica_Match",
+			Node: nodeWithZoneLabels,
+			ZonesWithNodes:      "zoneW,zoneX",
+			ReplicaCount:        2,
+			Reject:              false,
+			ExpectSpecificZones: true,
+			ExpectedZones:       "zoneW,zoneX",
```

Comment:
I know ExpectedZone is redundant when ExpectSpecificZones is true, but may be good to explicitly call out anyway.